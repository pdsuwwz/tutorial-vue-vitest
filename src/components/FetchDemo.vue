<template>
  <div>
    <h2>{{ message }}</h2>
    <ul>
      <li v-for="item in items" :key="item.id" @click="selectItem(item.id)">
        {{ item.name }}
      </li>
    </ul>
    <button @click="fetchData">Fetch Data</button>
  </div>
</template>

<script>
import { ref, reactive, defineComponent } from 'vue'

export default defineComponent({
  setup() {
    const message = ref('Hello')
    const items = reactive([])

    const fetchData = async () => {
      const response = await fetch('https://api.example.com/data')
      const data = await response.json()
      items.push(...data)
    }

    const selectItem = (id) => {
      message.value = `Selected item ${id}`
    }

    return {
      message,
      items,
      fetchData,
      selectItem
    }
  }
})
</script>
