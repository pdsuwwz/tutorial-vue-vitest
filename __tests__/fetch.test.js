import { describe, vi, beforeEach, it, expect } from 'vitest'

import { mount, flushPromises } from '@vue/test-utils'
import FetchDemo from './src/components/FetchDemo.vue'


describe('Test FetchDemo', () => {
  
  // beforeEach() 所注册的回调会在当前 describe 上下文中的每个测试运行之前被调用一次
  beforeEach(() => {
    // 定义全局方法，使每个 it 都能用到此方法
    global.fetch = vi.fn().mockResolvedValue({
      json: vi.fn().mockResolvedValue([
        { id: 1, name: 'Item 1' },
        { id: 2, name: 'Item 2' },
        { id: 3, name: 'Item 3' }
      ])
    })
  })
  
  it('组件正常渲染', () => {
    const wrapper = mount(FetchDemo)
    
    expect(wrapper.find('h2').text()).toBe('Hello')
    expect(wrapper.findAll('li')).toHaveLength(0)
    expect(wrapper.find('button').text()).toBe('Fetch Data')
  })

  it('fetch 数据 + 渲染列表', async () => {
    const wrapper = mount(FetchDemo)

    await wrapper.find('button').trigger('click')
    await wrapper.vm.$nextTick()

    
    expect(global.fetch).toHaveBeenCalledWith('https://api.example.com/data')

    
    await flushPromises()
    
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findAll('li')).toHaveLength(3)
    expect(wrapper.findAll('li')[0].text()).toBe('Item 1')
    expect(wrapper.findAll('li')[1].text()).toBe('Item 2')
    expect(wrapper.findAll('li')[2].text()).toBe('Item 3')
  })

  it('item 点击测试', async () => {
    const wrapper = mount(FetchDemo)

    await wrapper.find('button').trigger('click')
    await wrapper.find('button').trigger('click')
    await wrapper.find('button').trigger('click')
    await wrapper.find('button').trigger('click')

    expect(global.fetch).toHaveBeenCalledWith('https://api.example.com/data')
    expect(global.fetch).toHaveBeenCalledTimes(4)


    await flushPromises()
    expect(wrapper.html()).toMatchSnapshot()


    await wrapper.findAll('li')[1].trigger('click')
    await wrapper.vm.$nextTick()

    expect(wrapper.find('h2').text()).toBe('Selected item 2')
  })
})
