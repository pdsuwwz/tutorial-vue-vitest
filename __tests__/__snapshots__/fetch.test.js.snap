// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test FetchDemo > fetch 数据 + 渲染列表 1`] = `
"<div>
  <h2>Hello</h2>
  <ul>
    <li>Item 1</li>
    <li>Item 2</li>
    <li>Item 3</li>
  </ul><button>Fetch Data</button>
</div>"
`;

exports[`Test FetchDemo > item 点击测试 1`] = `
"<div>
  <h2>Hello</h2>
  <ul>
    <li>Item 1</li>
    <li>Item 2</li>
    <li>Item 3</li>
    <li>Item 1</li>
    <li>Item 2</li>
    <li>Item 3</li>
    <li>Item 1</li>
    <li>Item 2</li>
    <li>Item 3</li>
    <li>Item 1</li>
    <li>Item 2</li>
    <li>Item 3</li>
  </ul><button>Fetch Data</button>
</div>"
`;
